### PR TITLE
8322735: C2: minor improvements of bubble sort used in SuperWord::packset_sort

### PR DESCRIPTION
**Pull request by D-D-H**
A minor improvement could be made for bubble sort in SuperWord::packset_sort to reduce the comparison count in bad cases.

See https://en.wikipedia.org/wiki/Bubble_sort

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322735](https://bugs.openjdk.org/browse/JDK-8322735): C2: minor improvements of bubble sort used in SuperWord::packset_sort (**Enhancement** - P5)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [ba53ed56](https://git.openjdk.org/jdk/pull/17190/files/ba53ed56f9d480138f6bf4ddad9b48ca214d9a7d)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17190/head:pull/17190` \
`$ git checkout pull/17190`

Update a local copy of the PR: \
`$ git checkout pull/17190` \
`$ git pull https://git.openjdk.org/jdk.git pull/17190/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17190`

View PR using the GUI difftool: \
`$ git pr show -t 17190`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17190.diff">https://git.openjdk.org/jdk/pull/17190.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17190#issuecomment-1869031503)

### CODE REVIEW

**Comment by eme64** [comment]
@D-D-H can you explain what this improves?

**Comment by D-D-H** [comment]
> And a more fundamental question: Why do we need this improvement? Do you see any timing bottleneck and improvement? And what is faster: bubbling up or down?
> 
> And do you know why we sort at all in `extend_packlist` and why we do it again and again?

Sorry, I don't know the theory or implementation of `superword`. (I hope to grasp it someday...)

I just found it when browsing the code. This change is trivial; if you think it's unnecessary, I'm fine with closing it.

**Comment by eme64** [comment]
Ok. This change is fine with me. I was just curious what was your motivation. I may completely redo this code once I remove the alignment constraints (here used for sorting), but that will have to be decided in a few months.

Please do the renaming, and then I can run testing and give you my approval.

**Comment by D-D-H** [comment]
> Please do the renaming, and then I can run testing and give you my approval.

Updated. Thanks.

**Comment by eme64** [comment]
@D-D-H nice. Testing running.

**Review by eme64 — APPROVED** [review]
@D-D-H testing passed. Looks good. Thanks for the change!

**Comment by D-D-H** [comment]
@eme64 
Thank you!  Do I need a second review?

**Comment by eme64** [comment]
Yes, I think that would be preferrable, even though this is not a very complicated fix.

**Comment by D-D-H** [comment]
> Yes, I think that would be preferrable, even though this is not a very complicated fix.

Okay. Thanks.

File: src/hotspot/share/opto/superword.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/superword.cpp:3529]: So we not using `i+1` here because all previous values should be < than `i`'s
Right?

**Review by vnkozlov — COMMENTED** [review]
Clever.

File: src/hotspot/share/opto/superword.cpp
Comment (D-D-H) [reply src/hotspot/share/opto/superword.cpp]: Good catch. Deleted.

File: src/hotspot/share/opto/superword.cpp
Comment (D-D-H) [reply src/hotspot/share/opto/superword.cpp:3528]: Yes.
The last `i`'s value is > previous values and values between `i` and end are already sorted.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by D-D-H** [comment]
/integrate

Thanks for the review.

**Pull request closed without merging**
